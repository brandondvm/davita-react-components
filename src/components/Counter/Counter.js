import React, { useState } from 'react';
import PropTypes from 'prop-types';

const Counter = ({ id, value = 0, name, label = '', labelClass = 'yellow', isReadOnly, register, onCounterChange }) => {
  const [ counter, setCounter] = useState(value);

  function counterIncrement() {
    const newValue = Number(counter) + 1;
    setCounter(newValue);
    onCounterChange && onCounterChange(name, newValue);
  };

  function counterDecrement() {
    const newValue = Number(counter) - 1;
    if (newValue >= 0) {
      setCounter(newValue);
      onCounterChange && onCounterChange(name, newValue);
    }    
  };

  function onValueChange(value) {
    value ? setCounter(Number(value)) : setCounter(value);
  };

  return (
    <div className="counter-component">      
      {
        label && (
          <div className="counter-component_label-wrapper">
            <div className={`counter-component_label_circle ${ labelClass }`}></div>
            <p>{ label }</p>
          </div>
        )        
      }
      {
        isReadOnly ? <input className="counter-component_readonly" type="text" ref={register} name={name} readOnly /> : (
          <div className="counter-component_input-wrapper">
            <div className={`counter-component_btn counter-component_btn-decrement ${counter <= 0 ? 'disabled' : ''}`} onClick={ counterDecrement }></div>
            <input id={id} className="counter-component_input" type="number" min="0" value={!register ? counter : undefined} ref={register} name={ name } onChange={ ({ target }) => onValueChange(target.value) } />
            <div className="counter-component_btn counter-component_btn-increment" onClick={ counterIncrement }></div>
          </div>
        )
      }
    </div>
  );
};

Counter.propTypes = {
  id: PropTypes.string,
  isReadOnly: PropTypes.bool,
  label: PropTypes.string,
  labelClass: PropTypes.string,
  name: PropTypes.string,
  register: PropTypes.func,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default Counter;
