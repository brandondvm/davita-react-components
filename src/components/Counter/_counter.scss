.counter-component {
  align-items: center;
  display: flex;
  height: 32px;
  justify-content: space-between;

  input[type="number"]::-webkit-outer-spin-button,
  input[type="number"]::-webkit-inner-spin-button {
      -webkit-appearance: none;
      margin: 0;
  }
  input[type="number"] {
      -moz-appearance: textfield;
  }

  &_label-wrapper {
    align-items: center;
    display: flex;    
    p {
      color: $onyx;
      font-size: 14px;
      margin: 0;
    }
  }

  &_readonly {
    border: 0;
    color: $onyx;
    font-size: 14px;
    margin: 0;
    text-align: right;
    width: 40px;  

    &:focus {
      outline: none;
    }
  }

  &_label_circle {    
    border-radius: 100%;
    display: block;
    height: 16px;
    margin-right: 10px;
    width: 16px;
    &.yellow {
      background-color: $orange;
    }
    &.green {
      background-color: $lightgreen;
    }
    &.red {
      background-color: $red;
    }
    &.purple {
      background-color: $purple;
    }
  }

  &_input-wrapper {
    align-items: center;
    display: flex;    
    max-width: 96px;
  }

  &_btn {
    align-items: center;
    background-color: $darkblue;
    border-radius: 30px;
    color: $white;
    display: flex;
    height: 24px;
    justify-content: center;
    position: relative;
    width: 24px;

    &::before {
      border: 1px solid $white;
      content: '';
      display: block;
      height: 0;
      width: 12px;
    }

    &.disabled {
      background-color: $anotherpalegray;
    }
  }

  &_btn-decrement { 
    margin-right: 8px;

    &.disabled {
      background-color: $anotherpalegray;
      &::after {
        border: 1px solid $darkgray;
      }
      &::before {
        border: 1px solid $darkgray;
      }
    }
  }

  &_btn-increment {
    margin-left: 8px;
    &::after {
      border: 1px solid $white;
      content: '';
      display: block;
      height: 12px;
      left: 0;
      margin: 0 auto;
      position: absolute;
      right: 0;
      width: 0;
    }
  }

  &_input {
    border: 1px solid $darkgray;
    border-radius: 4px;
    font-family: $regular-font;
    font-size: 15px;
    height: 32px;
    width: 32px;
    text-align: center;
  }
}
