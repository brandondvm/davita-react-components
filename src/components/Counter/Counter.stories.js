import React from 'react';
import Counter from './Counter';

export default {
  title: 'Counter',
  component: Counter,
};

const Template = (args) => <div className="storybook-wrapper"><Counter {...args} /></div>;

export const Default = Template.bind({});
Default.args = {
  label: 'Counter Label',
  value: 0
};

export const NoLabel = Template.bind({});
