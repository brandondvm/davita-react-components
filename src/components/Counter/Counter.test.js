import enzyme from 'enzyme';
import React from 'react';

import Counter from './Counter';

const mockProps = {
  value: 0,
};

function setup(passedProps = {}) {
  const updatedProps = {
    ...mockProps,
    ...passedProps,
  };

  return enzyme.shallow(<Counter { ...updatedProps } />);
}

describe('Counter Component', () => {
  test('Component renders correctly', () => {
    const wrapper = setup();
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  test('Component renders label value form props correctly', () => {
    const label = 'Checkbox label';
    const wrapper = setup({ label });
    const checkboxLabel = wrapper.find('.counter-component_label-wrapper p');
    expect(checkboxLabel.text()).toEqual(label);
    expect(checkboxLabel).toBeDefined();
  });

  test('Component renders input value form props correctly', () => {
    const wrapper = setup({ value: 5 });
    const inputValue = wrapper.find('input').prop('value');
    expect(inputValue).toEqual(5);
  });

  test('Component renders decrement button as disabled when value is 0', () => {
    const wrapper = setup();
    const buttonDisabled = wrapper.find('.disabled');
    expect(buttonDisabled.length).toEqual(1);
  });

  test('Component renders decrement button as enable when value is bigger than 0', () => {
    const wrapper = setup({ value: 5 });
    const buttonDisabled = wrapper.find('.disabled');
    expect(buttonDisabled.length).toEqual(0);
  });

  test('Component increments state value on plus button clicks', () => {
    const wrapper = setup();
    wrapper.find('.counter-component_btn-increment').simulate('click');
    expect(wrapper.find('input').prop('value')).toBe(1);
  });

  test('Component decrements state value on minus button clicks', () => {
    const wrapper = setup({ value: 5 });
    wrapper.find('.counter-component_btn-decrement').simulate('click');
    expect(wrapper.find('input').prop('value')).toBe(4);
  });

  test('Component does not decrement state value on minus button clicks, when value is 0', () => {
    const wrapper = setup({ value: 0 });
    wrapper.find('.counter-component_btn-decrement').simulate('click');
    expect(wrapper.find('input').prop('value')).toBe(0);
  });
});
