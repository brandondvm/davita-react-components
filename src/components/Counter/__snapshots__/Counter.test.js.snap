// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Component renders correctly 1`] = `
<div
  className="counter-component"
>
  
  <div
    className="counter-component_input-wrapper"
  >
    <div
      className="counter-component_btn counter-component_btn-decrement disabled"
      onClick={[Function]}
    />
    <input
      className="counter-component_input"
      min="0"
      onChange={[Function]}
      type="number"
      value={0}
    />
    <div
      className="counter-component_btn counter-component_btn-increment"
      onClick={[Function]}
    />
  </div>
</div>
`;
